### PR TITLE
VarNetBlock

### PR DESCRIPTION
**Pull request by mersad95zd**
Signed-off-by: mersad95zd <m.zalbagi@gmail.com>

Fixes # .

### Description
This is the main component of the e2e-VarNet model for MRI reconstruction.
A VarNetBlock is called multiple times to create an e2e-VarNet model.

### Status
**Work in progress**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/apps/reconstruction/networks/blocks/varnetblock.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        self.model = refinement_model
+        self.spatial_dims = spatial_dims
+        self.dc_weight = nn.Parameter(torch.ones(1))  # learned scalar as the multiplier of the DC block
+
+        buffer_shape = [1 for _ in range(spatial_dims + 3)]  # 3 denotes the batch, channel, and real/complex dimensions
+        self.register_buffer("zeros", torch.zeros(buffer_shape))
+
+    def soft_dc(self, x: Tensor, ref_kspace: Tensor, mask: Tensor) -> Tensor:
+        """
+        Applies data consistency to input x. Suppose x is some intermediate estimate of the kspace and ref_kspace
```
Comment (mersad95zd) [inline monai/apps/reconstruction/networks/blocks/varnetblock.py:43]: Similar to the previous PR on sensitivity model, here the function `soft_dc` is specific to this `VarNetBlock` so I didn't put it under utils.

File: monai/apps/reconstruction/networks/nets/utils.py [hunk 2]
```diff
@@ -131,3 +134,43 @@ def complex_normalize(x: Tensor) -> Sequence:  # type: ignore
         )
         x = x.view(b, c, h, w, d)
         return (x - mean) / std, mean, std  # type: ignore
+
+
+def sens_reduce(x: Tensor, sens_maps: Tensor, spatial_dims: int = 2) -> Tensor:
+    """
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/nets/utils.py:140]: Could you help spell out the func name? Is it `reduce_sensitivity_map`?

Comment (Can-Zhao) [reply monai/apps/reconstruction/networks/nets/utils.py:140]: Or maybe `sum_sensitivity_map`?

File: monai/apps/reconstruction/networks/nets/utils.py [hunk 3]
```diff
@@ -131,3 +134,43 @@ def complex_normalize(x: Tensor) -> Sequence:  # type: ignore
...
+
+
+def sens_expand(x: Tensor, sens_maps: Tensor, spatial_dims: int = 2) -> Tensor:
+    """
+    Expands an image to its corresponding coil images based on the given sens_maps. Let's say there
+    are C coils. This function multiples image x with each coil sensitivity map in sens_maps and stacks
+    the resulting C coil images along the channel dimension which is reserved for coils.
+
+    Args:
+        x: 2D image (B,1,H,W,2) with the last dimension being 2 (for real/imaginary parts). 3D data will have
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/nets/utils.py:166]: If x is an image, would it be better to change it to `img`?

File: monai/apps/reconstruction/networks/nets/utils.py (hunk 2 above)
Comment (Can-Zhao) [reply monai/apps/reconstruction/networks/nets/utils.py:140]: If x is kspace signal, would it be better to call it kspace signal? I did not remember how you name kspace signal in your previous PRs.

File: monai/apps/reconstruction/networks/nets/utils.py [hunk 4]
```diff
@@ -131,3 +134,43 @@ def complex_normalize(x: Tensor) -> Sequence:  # type: ignore
...
+
+    Returns:
+        reduction of x to (B,1,H,W,2) for 2D data or (B,1,H,W,D,2) for 3D data.
+    """
+    x = ifftn_centered_t(x, spatial_dims=spatial_dims)  # inverse fourier transform
+    return complex_mul(x, complex_conj(sens_maps)).sum(dim=1, keepdim=True)  # type: ignore
+
+
+def sens_expand(x: Tensor, sens_maps: Tensor, spatial_dims: int = 2) -> Tensor:
+    """
```
Comment (Can-Zhao) [inline monai/apps/reconstruction/networks/nets/utils.py:160]: Same here. Could you help spell out the func name? Thanks!

File: monai/apps/reconstruction/networks/nets/utils.py (hunk 2 above)
Comment (mersad95zd) [reply monai/apps/reconstruction/networks/nets/utils.py:140]: yeah sure, maybe `sensitivity_map_reduce` and `sensitivity_map_expand`... The terms `reduce` and `expand` have intuitive counter meanings.

**Comment by Can-Zhao** [comment]
Looks good. I only have some minor comments

File: monai/apps/reconstruction/networks/nets/utils.py (hunk 3 above)
Comment (mersad95zd) [reply monai/apps/reconstruction/networks/nets/utils.py:166]: yeah, we could use that... and also `ksp` for the `sensitivity_map_reduce` function.

File: monai/apps/reconstruction/networks/nets/utils.py (hunk 2 above)
Comment (mersad95zd) [reply monai/apps/reconstruction/networks/nets/utils.py:140]: yeah kspace sounds good

**Review by wyli — APPROVED** [review]

**Pull request merged**